### DISCUSSION TITLE
How can I disable virtualization in datagrid

### DISCUSSION THREAD

**Discussion by hilk** [AvaloniaUI/Avalonia#12511] [Q&A]
I'm facing same issue with https://github.com/AvaloniaUI/Avalonia/issues/11989 in datagrid. so how can I disable virtualization? 

![bug](https://github.com/AvaloniaUI/Avalonia/assets/8522604/0dbb7a98-824a-460b-a9e0-6575f2a754b8)
sample project: https://github.com/hayriyedurgun/DataGridTest_v11

**Comment by timunie** [comment]
replace DataGrid.ItemsPanel with a StackPanel

**Reply by hilk** [reply]
There is no ItemsPanel in Avalonia.Controls.DataGrid/Themes/Fluent.xaml or Avalonia.Controls.DataGrid/Themes/Simple.xaml can you show me an example?

**Reply by timunie** [reply]
oh sorry, looks like DataGrid is much more complex here: https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Controls.DataGrid/Primitives/DataGridRowsPresenter.cs

Don't know if it's easier possible in TreeDataGrid?

**Reply by hilk** [reply]
TreeDataGrid is better in scrolling but there are still some issues. 
![treedatagrid](https://github.com/AvaloniaUI/Avalonia/assets/8522604/486a2d43-6c6d-48bd-941b-999b8df02827)
If this issue is related virtualization too, how can disable virtualization in TreeDataGrid? TreeDataGrid is using TreeDataGridRowPresenter.

**Comment by s-sarunas** [comment]
There's no direct way to disable virtualization in a DataGrid in Avalonia as far as I know.

As a workaround, you can place your DataGrid in a ScrollViewer with a vertical scroll bar and add a behavior to manually calculate the height of the DataGridRowsPresenter to fit all the rows:

```axaml
<ScrollViewer VerticalScrollBarVisibility="Auto">
  <DataGrid>
    <Interaction.Behaviors>
      <behaviors:DataGridVirtualizationDisablingBehavior />
    </Interaction.Behaviors>
  </DataGrid>
</ScrollViewer>
```

Behavior:

```csharp
public class DataGridVirtualizationDisablingBehavior : DisposingBehavior<DataGrid>
{
    protected override IDisposable OnAttachedOverride()
    {
        Guard.IsNotNull(AssociatedObject);
        Guard.IsOfType<DataGrid>(AssociatedObject);

        return AssociatedObject.AddDisposableHandler(DataGrid.TemplateAppliedEvent, OnTemplateApplied)!;
    }

    private void OnTemplateApplied(object? sender, TemplateAppliedEventArgs e)
    {
        var rowPresenter = AssociatedObject.FindDescendantOfType<DataGridRowsPresenter>();

        if (rowPresenter != null)
        {
            rowPresenter.Height = double.MaxValue;

            Dispatcher.UIThread.Post(() =>
            {
                double height = 0;

                foreach (var child in rowPresenter.Children)
                    height += child.DesiredSize.Height;

                rowPresenter.Height = height;
            }, DispatcherPriority.Loaded);
        }
    }
}
```

The height is calculated only once when the template for the DataGrid is applied. If you have row groups and/or row details you will need to recalculate the height every time those elements expand/collapse or elements inside change their height.
